docs/architecture/psa-thread-safety.md

```diff
@@ -293,8 +293,106 @@ There is currently no indication of when a slot is in the WRITING state. This on
+
+Potential ways for resolving this:
+1. Non-thread-safe drivers must not call the core
+2. Provide a new public API that drivers can safely call
+3. There is a whitelist of core APIs that drivers can call. Drivers providing entry points to these must not make a call to the core when handling these calls. (Drivers are still allowed to call any core API that can't have a driver entry point.)
+
+The first is too restrictive, the second is too expensive, the only viable option is the third.
+
+### Global Data
+
+PSA Crypto makes use of a `global_data` variable that will be accessible from multiple threads and needs to be protected. Any function accessing this variable (or its members) must take the corresponding lock first. Since `global_data` holds the RNG state, these will involve relatively expensive operations and therefore ideally `global_data` should be protected by its own, dedicated lock (different from the one protecting the key store).
+
+Note that this does not protect access to the RNG via `mbedtls_psa_get_random`, which is guaranteed to be thread-safe when `MBEDTLS_PSA_CRYPTO_EXTERNAL_RNG` is disabled. Still, doing so is conceptually simpler and we probably will want to remove the lower level mutex in the long run, since the corresponding interface will be removed from the public API. The two mutexes are different and are always taken in the same order, there is no risk of deadlock.
+
+The purpose of `MBEDTLS_PSA_CRYPTO_EXTERNAL_RNG` is very similar to the driver interface (and might even be added to it in the long run), therefore it makes sense to handle it the same way. In particular, we can use the `global_data` mutex to protect it as a default and when we implement the "thread_safe” property for drivers, we implement it for `MBEDTLS_PSA_CRYPTO_EXTERNAL_RNG` as well.
+
+### Implementation notes
+
+Since we only have simple mutexes, locking the same mutex from the same thread is a deadlock. Therefore functions taking the global mutex must not be called while holding the same mutex. Functions taking the mutex will document this fact and the implications.
+
+Releasing the mutex before a function call might introduce race conditions. Therefore might not be practical to take the mutex in low level access functions. If functions like that don't take the mutex, they need to rely on the caller to take it for them. These functions will document that the caller is required to hold the mutex.
+
+To avoid performance degradation, functions must not start expensive operations (eg. doing cryptography) while holding the mutex.
+
+## Strategy for 3.6
```

Comment:
The structure of the document is still unclear for someone who reads it in situ. It's currently:

> 1. Requirements
> 2. Resources to protect
> 3. Simple global lock strategy
> 4. Global lock excluding slot content (which has a section discussing condition variables)
> 5. Strategy for 3.6

There's no indication of what is implemented, what is planned to be implemented soon, what is planned to be implemented in the future, and what is ideas that are rejected. We don't need to clearly distinguish future plans vs rejected, but we do need to make it clear what is true vs what is being worked on vs the rest. I suggest the following:

* Add an introductory paragraph briefly discussing the current state (PSA calls are not thread-safe), what is planned for 3.6 (reference to the 3.6 strategy section), and mentioning that the rest of the document is design analysis.
* Create a “Design analysis” section. Write a short paragraph at the beginning stating that this section explores possible designs and does not reflect what is currently implemented. Make “Resources to protect”, “Simple global lock strategy” and “Global lock excluding slot content” subsections of that.